### PR TITLE
Fixes #2604

### LINKED ISSUE #2604: Support range-based charge queries from SMARTS
**Description:**
This would be `[+{1-}]`, etc.

### PR DESCRIPTION
**Pull request by greglandrum**
The one odd bit in here is that In order to have the negative charge version of this construct, `[-{1-3}]`, work I needed to add a new query type that handles negative formal charges. This is because the SMARTS parser doesn't actually recognize/handle negative numbers (and allowing it to do so would likely lead to a giant mess of shift/reduce problems).

### CODE REVIEW

File: Code/GraphMol/SmilesParse/smatest.cpp [hunk 1]
```diff
@@ -102,6 +102,8 @@ void testPass() {
     "[z{1-3}]",
     "[Z{1-3}]",
     "[2H,13C]",  // github #1719
+    "[+{0-3}]",
+    "[-{0-3}]",
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/smatest.cpp]: Could we try an or as well?

```
"[-{0-3},+{0-3}]"
```

File: Code/GraphMol/ChemReactions/Reaction.cpp [hunk 2]
```diff
@@ -211,11 +211,14 @@ bool ChemicalReaction::validate(unsigned int &numWarnings,
                qIter != query->endChildren(); ++qIter) {
             queries.push_back((*qIter).get());
           }
-          if (query->getDescription() == "AtomFormalCharge") {
+          if (query->getDescription() == "AtomFormalCharge" ||
+              query->getDescription() == "AtomNegativeFormalCharge") {
             int qval;
+            int neg =
+                query->getDescription() == "NegativeFormalCharge" ? -1 : 1;
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Reaction.cpp]: Shouldn't this be == "AtomNegativeFormalCharge"?  I can't imagine that this should ever be true.

File: Code/GraphMol/ChemReactions/Reaction.cpp [hunk 3]
```diff
@@ -224,9 +227,12 @@ bool ChemicalReaction::validate(unsigned int &numWarnings,
               }
               numWarnings++;
             } else {
+              int neg =
+                  query->getDescription() == "NegativeFormalCharge" ? -1 : 1;
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Reaction.cpp]: Here as well.   NegativeFormalCharge should be AtomNegativeFormalCharge?

File: Code/GraphMol/ChemReactions/Reaction.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/ChemReactions/Reaction.cpp]: Yep. Will fix it

Comment (greglandrum) [reply Code/GraphMol/ChemReactions/Reaction.cpp]: Yep. Will fix it

File: Code/GraphMol/SmilesParse/smatest.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/SmilesParse/smatest.cpp]: This is just checking that the syntax is ok and that the thing can be serialized/deserialized. We could add a functional test with or; do you have some reason to think that it’s likely to be a problem?

Comment (bp-kelley) [reply Code/GraphMol/SmilesParse/smatest.cpp]: Nope, just curious.

**Comment by bp-kelley** [comment]
For the last commit, I noticed that the tests didn't change.  How was this passing before or do we need a test that executes this changed code?

**Comment by greglandrum** [comment]
> For the last commit, I noticed that the tests didn't change. How was this passing before or do we need a test that executes this changed code?

Testing that would require adding a reaction with that primitive. Probably worth doing

File: Code/GraphMol/SmilesParse/smatest.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/SmilesParse/smatest.cpp]: what's hilarious is that `"[-{0-3},+{0-3}]"` actually *is* an error with the current SMARTS parser.

Comment (greglandrum) [reply Code/GraphMol/SmilesParse/smatest.cpp]: actually, combining any two range queries on a single atom leads to an error. Lovely

**Comment by greglandrum** [comment]
> > For the last commit, I noticed that the tests didn't change. How was this passing before or do we need a test that executes this changed code?
> 
> Testing that would require adding a reaction with that primitive. Probably worth doing

The test is in.

**Pull request merged**